lib/charms/operator_libs_linux/v0/apt.py

```diff
@@ -1314,48 +1422,309 @@ def _parse(line: str, filename: str) -> DebianRepository:
         else:
             raise InvalidSourceError("An invalid sources line was found in %s!", filename)
 
-    def add(self, repo: DebianRepository, default_filename: Optional[bool] = False) -> None:
-        """Add a new repository to the system.
+    def add(  # noqa: D417  # undocumented-param: default_filename intentionally undocumented
+        self, repo: DebianRepository, default_filename: Optional[bool] = False
+    ) -> None:
+        """Add a new repository to the system using add-apt-repository.
 
         Args:
-          repo: a `DebianRepository` object
-          default_filename: an (Optional) filename if the default is not desirable
+            repo: a DebianRepository object where repo.enabled is True
+        raises:
+            ValueError: if repo.enabled is False
+            CalledProcessError: if there's an error running apt-add-repository
+
+        WARNING: Does not associate the repository with a signing key.
+        Use `import_key` to add a signing key globally.
+
+        WARNING: Don't forget to call `apt.update` before installing any packages!
+        Or call `apt.add_package` with `update_cache=True`.
+
+        WARNING: the default_filename keyword argument is provided for backwards compatibility
+        only. It is not used, and was not used in the previous revision of this library.
         """
-        new_filename = "{}-{}.list".format(
-            DebianRepository.prefix_from_uri(repo.uri), repo.release.replace("/", "-")
+        if not repo.enabled:
+            raise ValueError("{repo}.enabled is {value}".format(repo=repo, value=repo.enabled))
+        _add_repository(repo)
+        self._repository_map[_repo_to_identifier(repo)] = repo
```

Comment:
Raising a `ValueError` here is technically a breaking change.

Previously, adding a disabled repository would write it to disk in commented out form. I don't think we can do that since we call `add-apt-repository` now.

It's just not clear what effect a user should expect from adding a disabled repository, hence raising an error, but for compatibility maybe it would be better to `logger.warning` and return early instead.

Reply:
Likely doesn't matter much, but I'd probably go with the warning+early return too.